File: src/slipwai/project/flags.py
"""The flag half of a project with somewhere to deploy: the reader, the gate, and the release decision.

Under a production target every commit that passes `verify` on `main` is applied to production, so a flag
is the only thing standing between a merge and a customer — which is what the constitution's
`build-once-deploy-is-not-release` requires and what MinimumCD means by trunk being releasable at every
commit. That makes the flag a piece of the *skeleton*, not a convention to describe: a backend gets a
reader in the same way it gets a health check.

One module rather than a row in four, because the parts only make sense together. `FLAG_READERS` says
where each backend's reader comes from and where it lands; `flag_reader` is what a backend's
`service_files` merges in; the same table spells the path for the prose in `AGENTS.md`, the ladder stage in
`commands/drive.md` and the release section of `docs/deployment.md`, so the file, the rule that points at
it, and the page that explains it cannot disagree. Nothing here is emitted under `--target none`: a project
with nowhere to deploy has no unsafe push to make, and `flags.auto.tfvars` — the one place a flag is
declared — is the target's file.
"""
from __future__ import annotations

from dataclasses import dataclass

from ..assets import LANGUAGE_ROOT, asset_tree
from ..catalog import CATALOG
from ..services import App, services_of
from ..targets import managed

# What prose calls a project's own package, the way `backends.event_store_directory` spells it: the
# directory is named after the project, so a document cannot name it and stay true for every project.
PACKAGE = "<package>"


@dataclass(frozen=True)
class FlagReader:
    """One backend's flag reader: where it is committed, where it lands, and how a slice asks it.

    `source` and `tests` are in the template spelling — `delivery_starter`, `deliverystarter` — because
    that is how they arrive, and each backend's own `name_service` renames them after the project along
    with everything else in the service. `spoken` turns either into the spelling a document has to use.
    """

    tree: str
    source: str
    tests: str
    call: str


# Keyed by backend. Both Java backends read one tree, for the reason they share `java/build/` and every
# `../java/` source in `service_layouts.py`: this class names no framework type — no `@ConfigProperty`, no
# `@Value` — so a second copy would have nothing to say differently and could only drift.
FLAG_READERS: dict[str, FlagReader] = {
    "typescript": FlagReader(
        tree="typescript/flags",
        source="src/flags.ts",
        tests="tests/flags.test.ts",
        call="flagEnabled('checkout-v2')",
    ),
    "python": FlagReader(
        tree="python/flags",
        source="src/delivery_starter/flags.py",
        tests="tests/test_flags.py",
        call='flag_enabled("checkout-v2")',
    ),
    "go": FlagReader(
        tree="go/flags",
        source="flags/flags.go",
        tests="flags/flags_test.go",
        call='flags.Enabled("checkout-v2")',
    ),
    "rust": FlagReader(
        tree="rust/flags",
        source="src/flags.rs",
        tests="src/flags.rs",
        call='flags::enabled("checkout-v2")',
    ),
    "java-quarkus": FlagReader(
        tree="java/flags",
        source="src/main/java/com/example/deliverystarter/flags/Flags.java",
        tests="src/test/java/com/example/deliverystarter/flags/FlagsTest.java",
        call='Flags.enabled("checkout-v2")',
    ),
    "java-spring": FlagReader(
        tree="java/flags",
        source="src/main/java/com/example/deliverystarter/flags/Flags.java",
        tests="src/test/java/com/example/deliverystarter/flags/FlagsTest.java",
        call='Flags.enabled("checkout-v2")',
    ),
}


def flag_reader(target: str, backend: str) -> dict[str, str]:
    """The reader and its own tests, keyed relative to a service — and nothing at all under `none`.

    Merged by each backend's `service_files`, before `name_service` runs, so the package rename reaches
    these files the way it reaches every other one: the Python reader lands in the project's own package
    and the Java one declares the project's own package, with no second place that knows how either is
    spelled.
    """
    if not managed(CATALOG, target):
        return {}
    return asset_tree(LANGUAGE_ROOT / FLAG_READERS[backend].tree)


def spoken(relative: str) -> str:
    """A committed path as a document has to name it, with the project's own package left as `<package>`."""
    return relative.replace("delivery_starter", PACKAGE).replace("deliverystarter", PACKAGE)


def reader_paths(apps: list[App]) -> str:
    """Every service's reader, as prose names it — one entry per distinct path, in service order."""
    return ", ".join(
        dict.fromkeys(
            f"`{service.path}/{spoken(FLAG_READERS[service.backend].source)}`"
            for service in services_of(apps)
        )
    )


def reader_calls(apps: list[App]) -> str:
    """How a slice asks, in each language the project's services are written in."""
    return ", ".join(
        dict.fromkeys(f"`{FLAG_READERS[service.backend].call}`" for service in services_of(apps))
    )


def flag_gate_dependency(target: str) -> str:
    """`check-flags` in `verify`'s dependency list, under a target and nothing otherwise.

    A flag is declared in `infra/service/flags.auto.tfvars`, which is the target's own file, so under
    `none` there is nowhere to declare one, nothing to read and no unsafe push to hold back. A gate that
    always passed would be worse than its absence: it would read as coverage this project does not have.
    """
    return " check-flags" if managed(CATALOG, target) else ""


def flag_gate(target: str) -> str:
    """The `check-flags` recipe, for a project that has somewhere a flag could be declared.

    Part of `verify` rather than of the production section beside `make flag`: it reads files and needs no
    account, so it belongs with the other four static gates and runs on every commit — which is the only
    point at which "this flag is read nowhere" is cheap to hear.
    """
    if not managed(CATALOG, target):
        return ""
    return """check-flags: ## Fail when a flag is declared and never read, read and never declared, seeded on, or tested on one path
\tpython3 scripts/check-flags.py
"""


def flag_stage(apps: list[App]) -> str:
    """The release-constraint stage of `commands/drive.md`'s ladder, for a project with a target.

    Between the slice's gaps review and its plan, because the plan is where the decision is written down
    and because deciding it later means deciding it after the code exists — which is when "it is only a
    small change" starts to sound reasonable.
    """
    return f"""**Release constraint** — the slice's flag is decided and written down in the plan: which
   releasable capability the slice belongs to, the flag key that holds it, and the fact that the flag is
   declared in `infra/service/flags.auto.tfvars` seeded `off` in the same change as the code that reads
   it. A flag covers a **releasable capability**, which is usually more than one slice: a capability that
   is not coherent to an actor until three slices land wants one flag held off across all three, not three
   flags — and one flag stretched over unrelated capabilities is the opposite failure, because they can
   then only be released together. So state which of the three this slice is, and why:
   it **continues** a capability that already has a flag and reuses that key; it **opens** one and needs a
   new key; or it is **releasable on merge and needs no flag**, because everything it adds is coherent and
   safe to an actor the moment it lands — a read-only path that shows only what the actor may already see
   is the usual case, and a change with no actor-visible behaviour at all is the other.
   The third answer is a decision like the other two and is written down like them: name it in the plan
   with its reason. An absent flag and an undecided flag look identical afterwards, and only one of them is
   safe — so a slice with no key says so in a sentence rather than saying nothing, and `AGENTS.md` asks for
   that same sentence again in the turn that pushes. **Recommend the answer with its reason rather than
   asking an open question**, and ask the user only where it is genuinely undecidable which capability the
   slice completes, or whether what it exposes is already the actor's to see — those are product questions
   and not defaults. Never add a second flag for a capability that already has one, and never gate a slice
   for the look of it: a flag over something already released is a code path kept alive for nobody.
   Where there is a flag, it is read **only** through {reader_paths(apps)}, asked by key
   ({reader_calls(apps)}) — never by naming `FLAG_<KEY>` at the point of use, which loses the transform,
   the single spelling and the seam that lets a test drive both paths. `make check-flags` refuses a key
   declared and never read, a key read and never declared, a new key seeded anything but `off`, a key
   whose tests only drive one path, and a read that goes round the reader."""


def release_notes(apps: list[App]) -> str:
    """The release half of `docs/deployment.md`: what a flag covers, which undo button, and what a flip is.

    Here rather than in `production_docs.py` for the reason the rest of this module is here: every sentence
    below is about the same three files — the reader, `flags.tf` and `flags.auto.tfvars` — and the page that
    explains them should not be able to drift from the rule in `AGENTS.md` that points at them.
    """
    return f"""
## Releasing, and undoing a release

Deploying and releasing are two decisions, and everything above this line is the first one. A commit that
passes `verify` on `main` is applied to production with nobody in between, so what makes that safe is that
unfinished work arrives **dark**: merged, deployed, and gated by a flag that is off.

A flag is declared in `infra/service/flags.auto.tfvars` under the service that reads it, seeded `off`, in
the same change as the code reading it. It reaches that service as `FLAG_<KEY>`, and the only place this
side reads one is
{reader_paths(apps)}.
Only place is meant literally: ask by key and the reader derives the variable, so the declaration and the
code cannot drift apart, an unset variable reads as off rather than as a typo, and a test can drive both
paths without touching the process's own environment.

Where the value comes *from* is one object inside that reader — the source — and it is keyed by the flag's
own key rather than by a variable name. It answers two questions: what this environment holds for one key,
and what it holds for all of them. Today there is one source and it reads this container's environment,
which is where the SSM parameter lands. A transport that is not an environment, an AppConfig agent beside
the container answered over loopback, is a second source and a one-line change to the reader's default: no
call site moves, no test moves, and this page's account of what a flip costs is the only other thing that
has to. The second question is there because something has to answer it for a browser app, which cannot
read this environment itself — and it is a snapshot rather than a subscription, because a transport polling
an agent can promise the values it last saw and nothing stronger. `make check-flags` runs inside `make verify` and
refuses a key declared and never read, a key read and never declared, a new key seeded anything but `off`,
a key whose tests only ever drive one path, and a read that names `FLAG_<KEY>` instead of calling the
reader.

### A flag covers a releasable capability, not a slice

"One flag per slice" is the wrong rule and it is the one teams reach for. A capability that is not coherent
to an actor until three slices land wants **one** flag, held off across all three: three flags mean two
states nobody designed, and the first one turned on shows a customer half a feature. The opposite failure
is one flag stretched over unrelated capabilities — they can then only be released together, and the flag
never gets deleted because something behind it is always unfinished. So the unit is the smallest thing
worth announcing, and the test of a key is whether turning it on is a sentence you would say to a customer.

### Two undo buttons, and they cost different things

| Something is wrong with | Press | What it costs |
|---|---|---|
| the release — it is broken however the flags are set | `.github/workflows/rollback.yml`, started by hand with the environment | re-applies the previous release's digests and its `index.html`, then smokes it. Every unrelated commit in that release goes backwards with it, and the schema does not move — which is why every migration is held to expand/contract |
| the flip — the release was fine until a flag went on | `make flag ENV=… KEY=… VALUE=off` | a rolling restart of one service, about two minutes. No build, no apply, no merge, and nothing unrelated moves |

Pressing the wrong one is the failure worth naming: **a rollback does not turn a flag off.** The parameter
is not in the release record, and `ignore_changes = [value]` means no apply ever puts it back — so a
rollback aimed at a bad flip returns good code to an earlier version and leaves the flag exactly as it was.
Ask which of the two rows you are in before pressing anything: did this break when the release went out, or
when somebody flipped something?

### The flip is the one production change outside the pipeline

`make flag` writes an SSM parameter and forces a new deployment. No `verify`, no review, no commit, and
nothing in the release record — which names a commit and its images. That is deliberate, and it is the one
place this project's own constitution is knowingly stretched: *One Path to Production* says that a thing
that can be changed in a running environment without a commit is not managed. The seed in
`flags.auto.tfvars` is managed; the live value is not, on purpose, because a release that needed a deploy
would not be a separate decision from one.

Two things make that affordable rather than merely true:

- **`make flags ENV=…` is where the answer lives**, and it prints more than the value: SSM keeps every
  parameter's version history, so it also prints who set each flag and when. Nothing extra is stored to
  make that work — a flip has an audit trail because the parameter store already kept one.
- **`make flag` smokes the environment afterwards.** The flip forces a new deployment, so the tasks that
  hold the new value have to come up and pass their health checks either way; waiting for that and then
  asking the environment whether it still answers costs the two minutes already being spent.

### Blue/green means both revisions read the same parameter

During the bake — five minutes in production — the previous release and the new one are both running, and
both resolve the same parameter. A flip in that window reaches both. A key the older revision has never
heard of is harmless, because nothing reads it; a key whose *meaning* changed between the two revisions is
not, and that is the compatibility rule landing on flags instead of on schemas. So a key is never
repurposed: retire it and introduce another, exactly as a column is expanded before it is contracted.

### What is not gated: owner, removal date, expiry

The constitution asks that every release flag have an owner, a removal date, tests on both paths, and an
expiry. `make check-flags` gates the third of those and nothing gates the rest, because on a boolean flag
they are hygiene rather than safety — a flag with no owner is untidy, a flag nobody reads is a feature that
never ships. Record them as a comment beside the key in `flags.auto.tfvars`, and when the capability is
released, delete the key, the branch it gated and the test of the off path in one change: a flag that
outlives its capability is a second code path being maintained for nobody.
"""


# What `docs/gates.md` says about the flag gate, beside what it says about the image half of `make ci`.
FLAG_GATE_NOTE = """
`make check-flags` is in `verify` for a reason of the same kind: every commit that passes it on `main`
reaches production, so a feature flag is the only thing between a merge and a customer. It refuses a flag
declared in `infra/service/flags.auto.tfvars` and read nowhere, a flag read and declared nowhere — which
exists in no environment and so can never be turned on — a new flag seeded anything but `off`, a flag
whose tests drive only one of its two paths, and a read that goes round the service's own reader to name
`FLAG_<KEY>` itself, which loses the key-to-variable transform and the seam both paths are tested through.
`docs/deployment.md` says what one flag covers, which undo
button a bad flip needs, and why the flip is the one production change made outside the pipeline.
"""


PUSH_CHECK = """### Before you push, name the release constraint

A push to `main` is a production deploy here: `verify` passes, `deploy.yml` applies staging, smokes it,
applies production and smokes that, with nobody in between. So the last thing before pushing is to say, in
the turn that pushes, what holds this change back from the actor — the flag key, and that it is `off` in
both environments.

**If nothing does — no flag, or a flag that is already on — stop and ask the user to confirm that this is a
release they want now.** Do not push and mention it afterwards, and do not decide on their behalf that the
change is small enough to be safe: there is no later gate at which somebody would catch it, which is the
whole reason the question is asked here. A refactor with no actor-visible behaviour is not a release and
needs no flag; say that it is one, rather than saying nothing.
"""

File: assets/languages/rust/flags/src/flags.rs
//! This service's feature flags, and the only place this side reads one.
//!
//! A flag is what makes merging and releasing two decisions. Every commit that passes `verify` on `main`
//! reaches production, so work that is not finished has to arrive there dark: the branch is merged, the flag
//! is off, and nobody outside sees it until somebody flips it. `.specify/memory/constitution.md` requires
//! exactly that, and this module is where the requirement stops being prose.
//!
//! # One flag, one name
//!
//! A flag is declared once, in `infra/service/flags.auto.tfvars`, under the service that reads it, with a key
//! in one spelling: `checkout-v2`. Ask by key and the declaration and the code cannot drift apart — which is
//! the whole reason a slice calls `flags::enabled("checkout-v2")` and never reaches for a variable name of its
//! own. A hand-derived variable is a typo waiting to happen, and a typo reads as absent, which reads as off.
//!
//! # Where a value comes from is one value, and it is not this function
//!
//! A [`Source`] answers two questions and no others: what this environment holds for one key, and what it
//! holds for all of them. Today [`default_source`] returns the process environment: the stack turns each key
//! into a parameter resolved into this container's environment as `FLAG_CHECKOUT_V2` — upper-cased, dashes to
//! underscores. [`variable`] is that transform and [`key`] its inverse, both written here and nowhere else. A
//! transport that is not an environment is a second `Source` and a one-line change to `default_source`, not a
//! change to any call site.
//!
//! # Off is the answer to every question this cannot answer
//!
//! A flag is on only for the exact value `on` — the spelling `make flag` writes. `off`, `true`, `1`, a value
//! that never arrived: all off. That is the safe direction, and it matters most between merging code that
//! reads a flag and the apply that creates its parameter: the feature starts quietly off rather than the
//! service failing to start.
//!
//! # The seam exists so both paths can be tested
//!
//! A test drives either path by passing a source — `fixed_source([("checkout-v2", "on")])` for the on path
//! and `fixed_source([])` for the one production runs while the flag is off. `make check-flags` holds every
//! declared flag to having both paths covered. `std::env` is read here and in no other module.
//!
//! Locally there is no parameter store: the flag is whatever this process's environment says, so
//! `FLAG_CHECKOUT_V2=on make dev` is the whole of it.

use std::collections::BTreeMap;

/// The only value that turns a flag on. Anything else, or nothing at all, gates its feature shut.
pub const ON: &str = "on";

/// What the stack gives every flag variable, and what [`key`] will answer to.
const PREFIX: &str = "FLAG_";

/// Where this service's flags come from, keyed by the flag's key and never by a variable name.
pub trait Source {
    /// This environment's raw value for one flag, or the empty string when it carries none.
    fn value(&self, key: &str) -> String;

    /// Every flag this source carries, by key, as of now — a snapshot, never a subscription. Only flags with a
    /// value appear: an absent flag is off, and saying so by omission is the answer `value` gives.
    fn snapshot(&self) -> BTreeMap<String, String>;
}

/// The environment variable a flag's key is read from: `checkout-v2` becomes `FLAG_CHECKOUT_V2`.
pub fn variable(key: &str) -> String {
    format!("{PREFIX}{}", key.replace('-', "_").to_uppercase())
}

/// The key a flag variable came from: `FLAG_CHECKOUT_V2` becomes `checkout-v2`, and a variable that is not a
/// flag's has none.
///
/// Exact only because a key may not contain an underscore — `check-flags.py` holds every declared key to
/// `[a-z0-9][a-z0-9-]*` — so every underscore in the variable came from a dash. Anchored on the prefix, which
/// is why `VITE_FLAG_CHECKOUT_V2`, the browser's spelling of the same flag, is not a flag variable here.
pub fn key(variable: &str) -> Option<String> {
    variable
        .strip_prefix(PREFIX)
        .map(|rest| rest.replace('_', "-").to_lowercase())
}

fn flags_of<'a>(variables: impl Iterator<Item = (&'a str, &'a str)>) -> BTreeMap<String, String> {
    variables
        .filter_map(|(name, value)| key(name).map(|flag| (flag, value.to_owned())))
        .collect()
}

struct ProcessEnvironment;

impl Source for ProcessEnvironment {
    fn value(&self, key: &str) -> String {
        std::env::var(variable(key)).unwrap_or_default()
    }

    fn snapshot(&self) -> BTreeMap<String, String> {
        let variables: Vec<(String, String)> = std::env::vars().collect();
        flags_of(
            variables
                .iter()
                .map(|(name, value)| (name.as_str(), value.as_str())),
        )
    }
}

struct EnvironmentSource(BTreeMap<String, String>);

impl Source for EnvironmentSource {
    fn value(&self, key: &str) -> String {
        self.0.get(&variable(key)).cloned().unwrap_or_default()
    }

    fn snapshot(&self) -> BTreeMap<String, String> {
        flags_of(
            self.0
                .iter()
                .map(|(name, value)| (name.as_str(), value.as_str())),
        )
    }
}

struct FixedSource(BTreeMap<String, String>);

impl Source for FixedSource {
    fn value(&self, key: &str) -> String {
        self.0.get(key).cloned().unwrap_or_default()
    }

    fn snapshot(&self) -> BTreeMap<String, String> {
        self.0.clone()
    }
}

/// A source over this process's own environment — the only one that knows a flag is carried under a
/// different name than the one it is declared with.
pub fn process_environment() -> impl Source {
    ProcessEnvironment
}

/// A source over the given variables. A key the map does not carry is off, exactly as an unset variable is.
pub fn environment_source<'a>(
    variables: impl IntoIterator<Item = (&'a str, &'a str)>,
) -> impl Source {
    EnvironmentSource(
        variables
            .into_iter()
            .map(|(name, value)| (name.to_owned(), value.to_owned()))
            .collect(),
    )
}

/// A source over flags held by key — the key itself, so a test never spells a variable name.
pub fn fixed_source<'a>(values: impl IntoIterator<Item = (&'a str, &'a str)>) -> impl Source {
    FixedSource(
        values
            .into_iter()
            .map(|(flag, value)| (flag.to_owned(), value.to_owned()))
            .collect(),
    )
}

/// Where this service's flags come from — the one line a change of transport is.
pub fn default_source() -> impl Source {
    process_environment()
}

/// Whether the named flag is on, read through this service's own source.
pub fn enabled(key: &str) -> bool {
    enabled_in(key, &default_source())
}

/// Whether the named flag is on in the given source, which is what a test drives both paths with.
pub fn enabled_in(key: &str, source: &impl Source) -> bool {
    source.value(key) == ON
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn variable_spells_a_flag_the_way_the_stack_does() {
        // The same transform as `FLAG_${upper(replace(flag.key, "-", "_"))}` in infra/service/flags.tf. If this
        // changes the flag stops arriving — and an absent flag reads as off, so nothing fails loudly.
        assert_eq!(variable("checkout-v2"), "FLAG_CHECKOUT_V2");
        assert_eq!(variable("publish-table"), "FLAG_PUBLISH_TABLE");
    }

    #[test]
    fn key_is_the_exact_inverse_of_variable() {
        for flag in ["checkout-v2", "publish-table", "a", "b2b-invoicing-v10"] {
            assert_eq!(key(&variable(flag)).as_deref(), Some(flag));
        }
    }

    #[test]
    fn a_variable_that_is_not_a_flag_has_no_key() {
        for name in ["DATABASE_URL", "PGSSLMODE", "VITE_FLAG_CHECKOUT_V2"] {
            assert_eq!(key(name), None, "{name}");
        }
    }

    #[test]
    fn an_environment_source_reads_a_key_under_the_name_the_stack_gives_it() {
        assert_eq!(
            environment_source([("FLAG_CHECKOUT_V2", "on")]).value("checkout-v2"),
            "on"
        );
        assert_eq!(environment_source([]).value("checkout-v2"), "");
    }

    #[test]
    fn an_environment_source_snapshots_every_flag_and_nothing_that_is_not_one() {
        let source = environment_source([
            ("FLAG_CHECKOUT_V2", "on"),
            ("FLAG_PUBLISH_TABLE", "off"),
            ("DATABASE_URL", "postgres://nope"),
            ("VITE_FLAG_CHECKOUT_V2", "on"),
        ]);
        let want: BTreeMap<String, String> = [("checkout-v2", "on"), ("publish-table", "off")]
            .map(|(k, v)| (k.to_owned(), v.to_owned()))
            .into();
        assert_eq!(source.snapshot(), want);
    }

    #[test]
    fn a_fixed_source_is_keyed_by_the_flag_key() {
        assert_eq!(
            fixed_source([("checkout-v2", "on")]).value("checkout-v2"),
            "on"
        );
        assert_eq!(fixed_source([]).value("checkout-v2"), "");
        let want: BTreeMap<String, String> = [("checkout-v2".to_owned(), "on".to_owned())].into();
        assert_eq!(fixed_source([("checkout-v2", "on")]).snapshot(), want);
    }

    #[test]
    fn the_default_source_is_the_transport_this_service_has() {
        // Asked for a key no project would declare, so somebody who exported a real flag locally does not fail
        // this suite by doing so.
        assert_eq!(default_source().value("no-flag-sets-this"), "");
    }

    #[test]
    fn the_process_environment_snapshot_carries_only_flags() {
        // Reads the real environment without writing to it: `set_var` is unsafe in edition 2024, and this
        // crate forbids unsafe code. Whatever flags this process happens to carry, nothing that is not one
        // may appear — `PATH` is always set and never a flag.
        let snapshot = process_environment().snapshot();
        assert!(!snapshot.contains_key("path"));
        assert!(snapshot.keys().all(|flag| !flag.contains('_')));
    }

    #[test]
    fn a_flag_is_on_only_for_the_value_make_flag_writes() {
        assert!(enabled_in(
            "checkout-v2",
            &fixed_source([("checkout-v2", "on")])
        ));
        assert!(!enabled_in(
            "checkout-v2",
            &fixed_source([("checkout-v2", "off")])
        ));
    }

    #[test]
    fn a_flag_nothing_set_is_off() {
        assert!(!enabled_in("checkout-v2", &fixed_source([])));
        assert!(!enabled("no-flag-sets-this"));
    }

    #[test]
    fn a_value_the_reader_does_not_understand_is_off() {
        for value in ["true", "1", "ON", "yes"] {
            assert!(
                !enabled_in("checkout-v2", &fixed_source([("checkout-v2", value)])),
                "{value}"
            );
        }
    }

    #[test]
    fn a_flag_reads_through_whichever_source_it_is_given() {
        assert!(enabled_in(
            "checkout-v2",
            &environment_source([("FLAG_CHECKOUT_V2", "on")])
        ));
        assert!(!enabled_in("checkout-v2", &environment_source([])));
    }
}
